### ISSUE TITLE
Improve `knitr.table.format` handling

### ISSUE THREAD

**Issue by salim-b** [haozhu233/kableExtra#561]
I really like the goal of kableExtra – complementing `knitr::kable()` with additional functionality to produce both complex HTML and LaTeX tables! And until [gt](https://github.com/rstudio/gt) is more matured (i.e. supports LaTeX output) I'd like to rely on kableExtra.

But kableExtra's current handling of the kable `format` and  knitr's corresponding global option `knitr.table.format` is suboptimal, i.e. totally inconsistent IMHO.

knitr [allows](https://bookdown.org/yihui/rmarkdown-cookbook/kable.html#kable-formats) to set the global option `knitr.table.format` either directly to a valid format string (like `html`) or to a function returning one of these strings conditionally (which is the most flexible/powerful way to use `knitr::kable()`).

But kableExtra currently

- **overwrites** the global option `knitr.table.format` on package load with a static value depending on `knitr::is_latex_output()` (`latex` or `html`) _unless_ another global option `kableExtra.auto_format` is set to `FALSE`.

   This in itself is a bad design choice – overwriting a global option introduced by _another_ package without at least informing the user about it.

   It's also suboptimal to just base the kable `format` on the return value of `knitr::is_latex_output()` at the time the kableExtra package is loaded.

- Some of kableExtra's functions which expose a `format` argument do look up the global option `knitr.table.format` if `format` is not explicitly provided. But they don't take the possibility into account that `knitr.table.format` could be set to a _function_ rather than a static value (in which case these functions just fail). Among these are e.g. the `footnote_marker_*()` and `cell_spec()` functions.

- At least one of kableExtra's functions exposing a `format` argument does _not_ look up the global option `knitr.table.format` if `format` is not explicitly provided: `kbl()`

  Only the function `kableExtra::kable()` which is reexported from knitr does properly respect the global option `knitr.table.format` (i.e. including being set to a function).

Therefore I propose

1. to introduce a new utility function along the lines of [`pal::knitr_table_format()`](https://pal.rpkg.dev/reference/knitr_table_format.html):

    ```r
    #' Determine current knitr table format
    #'
    #' Determines the current knitr table format based on the \R option
    #' [`knitr.table.format`](https://bookdown.org/yihui/rmarkdown-cookbook/kable.html#kable-formats) which can either be set directly to a valid format string or
    #' to a function returning one of these strings conditionally.
    #'
    #' This is basically a convenience wrapper to be able to access the current `knitr.table.format` in a hassle-free way, i.e. it provides the conditional logic to
    #' account for the possibility that `knitr.table.format` is set to a function rather than a format string.
    #'
    #' @param default The knitr table format to fall back to when the \R option `knitr.table.format` is not set. One of
    #'   `"html"` or `"latex"`.
    #'   
    #' See [knitr::kable()]'s `format` argument for details.
    #'
    #' @return A character scalar.
    #' @export
    knitr_table_format <- function(default = c("html", "latex")) {
      
      allowed_formats <- eval(formals()$default)
      
      result <- getOption("knitr.table.format",
                          default = rlang::arg_match(default))
      
      if (is.function(result)) result <- result()
      
      if (!(result %in% allowed_formats)) {
        allowed_formats_print <- paste0('"', paste0("`", allowed_formats, "`")), '"')
        if (result %in% c("pipe", "simple", "rst")) {
          rlang::abort(paste0('Global option `knitr.table.format` is set to an unsupported format: `"', result, '"`. kableExtra only supports ', allowed_formats_print))
        } else {
          rlang::abort(paste0("Invalid global option `knitr.table.format`. Must evaluate to one of ", allowed_formats_print)
        }
      }
      
      result
    }
    ```

    (Maybe my assumption in the code above that kableExtra only supports the `html` and `latex` formats is not quite correct.)

2. to change kableExtra's package load behaviour to set the global option `knitr.table.format` to

    ```r
    function() if(knitr::is_latex_output()) "latex" else "html"
    ```

    _if_ it isn't already set to another value.

`knitr_table_format()` described above could then be used internally by all of kableExtra's functions requiring a `format` argument. This would be at least as user-friendly as kableExtra's current behaviour but also be _more dynamic_ by allowing to knit to both HTML and LaTeX output in the same R session. 

@haozhu233 What do you think about this? I'd be happy to contribute all the necessary changes.

**Comment by haozhu233** [comment]
Yeah, I agree that from a CS point of view, changing another package's global setting is a bad design. However, this design was made for a reason. For a very long time, this package relies solely on `knitr::kable` to generate tables, which by default returns markdown tables. I thought I explained well enough in both README and vignette at that time that you need to either specify `format` in `kable` or setup the global environment. However, I still receive numerous github issus, stackoverflow question or even direct emails asking me why kableExtra doesn't work on their (markdown) kable codes. That leads to the decision to change the global option of `kable` in `knitr`. I don't like this decision either and that's the reason why I added `kbl` in the last version, which wraps up `kable` so the old global option trick is not necessary. However, it still needs to be there because if I remove that it will break many people's code. 

Anyway, for your comments, 
1. Yeah, I agree that is a bad decision but if we change it, it will break a lot of existing codes can cause bad consequences. In addition, this is probably the most harmless way to make this code work `knitr::kable() %>% kableExtra::kable_styling()`, which I have seen too many times. 
2. Yeah, you are right. With the current settings, I think we should remove the `knitr.table.format` dependency from those functions. I prefer to see something similar with the current `kbl` function, which only returns latex when `knitr::is_latex_output()` is true. 
3. `kbl` was created to replace the old method of changing global option. I think we should not let it take the global option again. If the function is running under rmarkdown latex, it returns latex, otherwise it return html. It should be that simple.

**Comment by salim-b** [comment]
Thanks for your anwer.

> However, this design was made for a reason. For a very long time, this package relies solely on `knitr::kable` to generate tables, which by default returns markdown tables.

Sure, I totally understand the motivation behind kableExtra's current design.

> I thought I explained well enough in both README and vignette at that time that you need to either specify `format` in `kable` or setup the global environment.

I know and read the explanation in your README. Maybe I didn't express my concerns about kableExtra's current behaviour clearly enough: My main points of criticism refer to the facts that kableExtra currently

1. _overwrites_ the global option `knitr.table.format`, i.e. it sets its own value _regardless_ of the state of `knitr.table.format`, and it doesn't even print a message or something when it silently overwrites this option. This is in stark contrast to how global options are intended to be used in R. I think it would be fine if kableExtra would write to `knitr.table.format` _if and only if_ it is not already set to a valid format string or function.

2. does not support the global option `knitr.table.format` being set to a function. This constellation is e.g. possible when a user sets `knitr.table.format` to `function() if(knitr::is_latex_output()) "latex" else "html")` _after_ the kableExtra package has been loaded. This is unfortunate I think since setting `knitr.table.format` to the above function is the most powerful (i.e. dynamic) kable setup.

> However, I still receive numerous github issus, stackoverflow question or even direct emails asking me why kableExtra doesn't work on their (markdown) kable codes. That leads to the decision to change the global option of `kable` in `knitr`.

Yeah, I can only imagine the various user complaints. 🙈

If you carefully read my proposal: It also relies on setting `knitr.table.format`, but _only_ if it hasn't already been set to a valid format string or function by the user before.

> I don't like this decision either and that's the reason why I added `kbl` in the last version, which wraps up `kable` so the old global option trick is not necessary. However, it still needs to be there because if I remove that it will break many people's code.

Well, I think we should play nicely with knitr's current design. The global option `knitr.table.format` has been there for quite a while now and unless knitr itself retires it, I think kableExtra should honor it so that users get consistent results when mixing `knitr::kable()` and `kableExtra::kbl()` and other kableExtra functions in the same code. The semi-official [R Markdown Cookbook](https://bookdown.org/yihui/rmarkdown-cookbook/kable.html#kable-formats) explains this option well and advises how to set it to a function allowing for _dynamic_ table `format`s, so we could also just refer to that book in kableExtra's README etc.

---

In response to your _numbered_ comments:

1. I respectfully disagree. I _think_ my proposal would not brake any user's existing kableExtra code. And I _don't think_ that kableExtra's current design is "the most harmless way to make this code work `knitr::kable() %>% kableExtra::kable_styling()`".

2. I think a combination of "both worlds" would be most promising: _Respecting_ `knitr.table.format` as long as it's set to a valid value/function _and_ falling back to setting `knitr.table.format` to `function() if(knitr::is_latex_output()) "latex" else "html")` otherwise.

    Setting `knitr.table.format` to the above function rather than a format string has the **great advantage that you can successfully knit to both HTML and PDF/LaTeX _in the same R session_**. This is _not_ the case for kableExtra's current design.

3. I strongly think that _all_ of kableExtra's functions that require a `format` argument should behave exactly the same. This would be the case with my proposal.

    > If the function is running under rmarkdown latex, it returns latex, otherwise it return html. It should be that simple.

    This is exactly what I want to achieve (unless the user manually (and thus consciously we have to assume) set `knitr.table.format` to something valid).

---

Maybe I should prepare a PR with all the necessary changes to kableExtra so you can better see what I have in mind?

**Comment by vincentarelbundock** [comment]
@haozhu233 I know this is a complicated issue, but do you think the recent changes "fix" most of this?

**Comment by haozhu233** [comment]
Yeah, mostly. 

Here is a summarization of recent changes:

Starting from the upcoming version 1.4.0, we will stop our evil solution on manipulating knitr's global setting. Instead, we introduced a markdown table parser, which will parse the inputting markdown table back to a dataframe and resend it to the `kbl` function. In this way, all the auto formatting happens within this package. 

In terms of latex table specifically, we also turned on booktabs and longtable by default to mimic the default behavior of how pandoc handle markdown tables. As a result, most users won't even feel the change. 

The only part that gets impacted is the table labels or references set within the `kable` function through the `label` option. For this part, first, people can set labels using knitr chunk options. Also, people can easily overcome it by either using kbl or setting format in kable.
